<a href="https://colab.research.google.com/github/Redcoder815/Machine_learning_phitron/blob/main/35LogisticRegressonPractice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np

In [2]:
def sigmoid(z):
  return 1 / (1 + np.exp((-0.5*z)+0.2))

In [3]:
print(sigmoid(2))

0.6899744811276125


Given a single data point: actual y = 1, predicted probability ŷ = 0.3.  
Calculate the Log Loss for this single example.

In [4]:
y = 1
prob = 0.3

In [5]:
x = (y * np.log(prob)) + ((1 - y) * np.log(1 - prob))
print(-x)

1.2039728043259361


In [6]:
def sigmoid(z):
  return 1 / (1 + np.exp(-z))

In [7]:
def make_prediction(X, W, b, threshold=0.5):
  m = X.shape[0]
  prob_list = np.zeros((m,))

  for i in range(m):
    z = np.dot(W, X[i]) + b
    prob_list[i] = sigmoid(z)

  prediction = (prob_list > threshold)
  return prediction

In [8]:
def compute_cost(X, y, W, b):
  m = X.shape[0]
  total_cost = 0.0

  for i in range(m):
    z = np.dot(W, X[i]) + b
    prob = sigmoid(z)
    total_cost += (y[i] * np.log(prob)) + ((1 - y[i]) * np.log(1 - prob))

  return total_cost / (-m)

In [9]:
def calculate_gradient(X, y, W, b):
  m, n = X.shape
  dj_dw = np.zeros((n,))
  dj_db = 0.0

  for i in range(m):
    z = np.dot(W, X[i]) + b
    prob = sigmoid(z)
    error = prob - y[i]

    dj_db += error

    for j in range(n):
      dj_dw[j] += error * X[i, j]

  return dj_dw / m, dj_db / m

In [10]:
def gradient_descent(X, y, w_input, b_input, max_iter, alpha=0.01):
  w = w_input
  b = b_input
  cost_history = []

  for i in range(max_iter):
    dj_dw, dj_db = calculate_gradient(X, y, w, b)

    w = w - alpha * dj_dw
    b = b - alpha * dj_db

    if i % 100 == 0:
      cost = compute_cost(X, y, w, b)
      cost_history.append(cost)
      print(f"Iter {i}: Cost {cost:.4f}")

  return w, b, cost_history

You train a logistic regression model with weight w = 2.0  and bias b = -1.0.  

For a feature vector x = [1.5, 0.5],  

a) Compute z = w·x + b  

b) Compute the predicted probability ŷ = σ(z).  

c) If actual y = 1, should the model update w to increase or

decrease? Brief some reasoning


In [11]:
x = np.array([1.5, 0.5])
w = np.array([2.0, 2.0])
b = -1.0
y_true = 1

In [13]:
z = np.dot(w, x) + b
print(z)
y_pred = sigmoid(z)
print(y_pred)

3.0
0.9525741268224334


c) Weight Update Analysis:
   - Target (y) = 1, Prediction (y_hat) = 0.9526
   - Error (y - y_hat) = 0.0474 (Positive error)
   - Since features x are positive, weights must increase to push y_hat closer to 1.


A logistic regression model gives the following probabilities for 5 test samples:  
[0.1, 0.4, 0.6, 0.9, 0.45`  
Actual labels: [0, 0, 1, 1, 0]  
Using threshold = 0.5:  
a) Find predicted classes.  
b) Calculate accuracy.  
c) How many false positives and false negatives?

In [14]:
probabilities = np.array([0.1, 0.4, 0.6, 0.9, 0.45])
actual_labels = np.array([0, 0, 1, 1, 0])
threshold = 0.5

In [15]:
predicted_class = (probabilities >= threshold).astype(int)
print(predicted_class)

[0 0 1 1 0]


In [16]:
correct_prediction = (predicted_class == actual_labels)
accuracy = np.mean(correct_prediction)
print(f'Accuracy: {accuracy * 100:.0f}%')

Accuracy: 100%


In [17]:
false_positives = np.sum((predicted_class == 1) & (actual_labels == 0))
false_negetives = np.sum((predicted_class == 0) & (actual_labels == 1))

print(false_positives)
print(false_negetives)

0
0


A logistic regression model is trained on a dataset with 2 features.  
The learned parameters are: w = [1.2, -0.8], b = 0.5.  
a) Write the equation of the decision boundary.  
b) A new point has x1 = 2.0, x2 = 1.5. Which class is predicted?  
c) How far (in terms of z value) is this point from the decision boundary?


In [18]:
w = np.array([1.2, -0.8])
b = 0.5
x_new = np.array([2.0, 1.5])

In [19]:
# Equation form: w1*x1 + w2*x2 + b = 0
print(f"a) Decision Boundary Equation:")
print(f"   {w[0]:.1f}*x1 + ({w[1]:.1f})*x2 + {b:.1f} = 0")

a) Decision Boundary Equation:
   1.2*x1 + (-0.8)*x2 + 0.5 = 0


In [20]:
# --- Part b) Predict class ---
# Calculate linear combination z
z_point = np.dot(w, x_new) + b

# Classification based on threshold z = 0 (equivalent to probability threshold 0.5)
predicted_class = 1 if z_point > 0 else 0
print(f"\nb) Predicted Class:")
print(f"   - Computed z value = {z_point:.1f}")
print(f"   - Predicted class  = {predicted_class}")


b) Predicted Class:
   - Computed z value = 1.7
   - Predicted class  = 1


In [21]:
# --- Part c) Distance in terms of z value ---
# Distance from the z=0 boundary is simply the magnitude of z at that point
z_distance = abs(z_point - 0.0)
print(f"\nc) Distance from boundary (in terms of z value):")
print(f"   - Distance = {z_distance:.1f} units")


c) Distance from boundary (in terms of z value):
   - Distance = 1.7 units


You are given a highly imbalanced dataset: 95% negative class (0) and 5% positive class (1).  
You train a logistic regression model and get 95% accuracy, but it predicts all samples as class 0.  
a) Why is accuracy misleading here?  
b) Which evaluation metric(s) would you use instead?  
c) How can you adjust the model to detect more positive samples?


a) Why is accuracy misleading here?Accuracy measures the total percentage of correct predictions out of all samples. Because 95% of the dataset belongs to the negative class (0), a completely useless model that simply guesses "0" for every single input will automatically achieve 95% accuracy. It completely fails to detect a single minority/positive sample (0% of class 1 detected), yet appears highly successful on paper.

b) Which evaluation metric(s) would you use instead?Instead of accuracy, you should look at metrics that isolate performance on the minority class:Recall (Sensitivity): Measures what fraction of the actual positive samples the model managed to catch. (In this problem, Recall is currently 0%).Precision: Measures what fraction of the model's positive predictions were actually correct.F1-Score: The harmonic mean of Precision and Recall. It provides a single balanced score for imbalanced data.ROC-AUC or PR-AUC: Measures the model's ability to rank positive instances higher than negative instances across all possible thresholds.

c) How can you adjust the model to detect more positive samples?To force the model to capture class 1 samples, you can use these common approaches:Lower the Classification Threshold: Instead of the default 0.5, classify samples as 1 if their predicted probability is higher than a lower threshold (e.g., 0.1 or 0.05).Class Weights (Cost-Sensitive Learning): Assign a higher loss penalty to misclassified positive samples during training (e.g., using class_weight='balanced' in scikit-learn).Resampling the Data: Use techniques like SMOTE (Synthetic Minority Over-sampling Technique) to create synthetic positive samples, or under-sample the majority class to balance the dataset before training.

In [22]:
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

# 1. Create a dummy imbalanced dataset (950 zeros, 50 ones)
np.random.seed(42)
X = np.random.randn(1000, 2)
# Ensure the labels are highly skewed (95% 0, 5% 1)
y = np.array([0] * 950 + [1] * 50)

# --- The Bad Baseline Model ---
# Standard logistic regression with no adjustments
model_naive = LogisticRegression()
model_naive.fit(X, y)
preds_naive = model_naive.predict(X)

print("=== Naive Model (Unadjusted) ===")
print(f"Accuracy: {accuracy_score(y, preds_naive) * 100:.1f}%")
print(f"Unique classes predicted: {np.unique(preds_naive)}")
print("\nClassification Report (Notice 0.00 recall for class 1):")
print(classification_report(y, preds_naive, zero_division=0))


# --- The Adjusted Model (Part c solution) ---
# Fixing it using class weights to penalize minority class mistakes
model_fixed = LogisticRegression(class_weight="balanced")
model_fixed.fit(X, y)
preds_fixed = model_fixed.predict(X)

print("\n=== Adjusted Model (With Balanced Class Weights) ===")
print(f"Accuracy: {accuracy_score(y, preds_fixed) * 100:.1f}%")
print("\nClassification Report (Now it successfully detects class 1!):")
print(classification_report(y, preds_fixed))

=== Naive Model (Unadjusted) ===
Accuracy: 95.0%
Unique classes predicted: [0]

Classification Report (Notice 0.00 recall for class 1):
              precision    recall  f1-score   support

           0       0.95      1.00      0.97       950
           1       0.00      0.00      0.00        50

    accuracy                           0.95      1000
   macro avg       0.47      0.50      0.49      1000
weighted avg       0.90      0.95      0.93      1000


=== Adjusted Model (With Balanced Class Weights) ===
Accuracy: 51.6%

Classification Report (Now it successfully detects class 1!):
              precision    recall  f1-score   support

           0       0.95      0.52      0.67       950
           1       0.05      0.50      0.09        50

    accuracy                           0.52      1000
   macro avg       0.50      0.51      0.38      1000
weighted avg       0.91      0.52      0.64      1000

